# Imports Required

In [1]:
import pandas as pd

# Importing Data

In [2]:
churn= pd. read_csv("Telco-Customer-Churn.csv")

### Understanding the Dataset

In [3]:
# Information on the dataset
churn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [4]:
#Checking if the dataset has any null values
display(churn.isnull().any())

customerID          False
gender              False
SeniorCitizen       False
Partner             False
Dependents          False
tenure              False
PhoneService        False
MultipleLines       False
InternetService     False
OnlineSecurity      False
OnlineBackup        False
DeviceProtection    False
TechSupport         False
StreamingTV         False
StreamingMovies     False
Contract            False
PaperlessBilling    False
PaymentMethod       False
MonthlyCharges      False
TotalCharges        False
Churn               False
dtype: bool

# Question 1:Demographic

### a. Which 2 demographics have proven to be the most popular with the telco brand?

In [5]:
#Demographics would be considered by Gender and If one is a senior Citizen or Not.
#In terms of Gender
churn['gender'].value_counts()

Male      3555
Female    3488
Name: gender, dtype: int64

In [6]:
#In terms of Senior Citizens where 0 would be a young citizen and 1 a senior citizen
churn['SeniorCitizen'].value_counts()

0    5901
1    1142
Name: SeniorCitizen, dtype: int64

In [7]:
#In terms of both Gender and Senior Citizens
churn.groupby(['gender', 'SeniorCitizen']).size()

gender  SeniorCitizen
Female  0                2920
        1                 568
Male    0                2981
        1                 574
dtype: int64

###### Conclusion
<br> 
*Males are more popular with the telco brand. 
<br> 
*Non-SeniorCitizens(Young Citizens) are more popular with the telco brand.
<br>
*Therefore, Males who are non-SeniorCitizens are most popular with the brand 

### b. If the marketing team wanted to increase customer retention, to whom should they target their advertisements to?

In [8]:
#To increase customer retention, It is cheaper and easier to advertise to already exisiting customers rather than new customers
#In this case it would be better to advertise for a group that has many existing customers
#To be more specific about the best group to advertize to, we use Gender,SeniorCitizen and Tenure(How many months the customer has used the brand)
churn.groupby(['gender','tenure','SeniorCitizen']).size()

gender  tenure  SeniorCitizen
Female  0       0                  5
        1       0                237
                1                 47
        2       0                105
                1                 25
                                ... 
Male    70      1                  7
        71      0                 68
                1                 18
        72      0                156
                1                 29
Length: 290, dtype: int64

###### Conclusion
<br>
->From above analysis, generally:
<br>
*In terms of Gender, Males would be best to advertise to.
<br>
*In terms of SeniorCitizen, the Non-SeniorCitizens(Young citizens) would be the best to advertise to.
<br>
*In terms of Tenure, those who have been using the brand for 1 month are the best to advertize to.
<br> <br>
->More Specifically:
<br>
*The best groups to target would be:
<br>
*Non-SeniorCitizen Males who have been using the brand for 1 month as this will reduce their churn.
<br>
*Non-SeniorCitizen Females who have been using the brand for 1 month as this will reduce their churn numbers.
<br> <br>
->Other groups to consider:
<br>
*Non-SeniorCitizen Males who have been using the brand for 72 months
<br>
*Non-SeniorCitizen Females who have been using the brand for 72 months
<br>
*Non-SeniorCitizen Females who have been using the brand for 2 months
<br>
*Non-SeniorCitizen Males who have been using the brand for 2 months

# Question 2: Services

### a. Which 3 services are contributing to a higher monthly charge and resulting to customers churning? (Use visualizations in explaining your analysis)

In [9]:
#For this case we will need to manipulate the dataset to be able to analyze
#For each service we will need to get the maximum monthly charge that is most repeated where customers actually churn
#Making a copy of the dataset in order to avoid permanent damage to the original one
services= churn
services.sample(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
6294  6615-ZGEDR    Male              0      No         No       1   
6201  6458-CYIDZ  Female              1      No         No       5   
4223  4029-HPFVY    Male              0      No         No       3   
6992  0218-QNVAS    Male              0     Yes        Yes      71   
6739  6994-KERXL    Male              0      No         No       4   
677   0822-GAVAP  Female              0      No         No       2   
6444  6302-JGYRJ    Male              0      No        Yes      31   
4870  5825-XJOCM    Male              0      No        Yes      48   
4702  6008-NAIXK    Male              1      No         No      54   
2361  5103-MHMHY  Female              0      No        Yes       1   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
6294          Yes                No              No  No internet service  ...   
6201          Yes                No     Fiber optic                  Yes  ...   
4223          Yes                No     Fiber optic                   No  ...   
6992          Yes               Yes     Fiber optic                   No  ...   
6739          Yes                No             DSL                   No  ...   
677            No  No phone service             DSL                   No  ...   
6444          Yes               Yes             DSL                   No  ...   
4870          Yes               Yes             DSL                  Yes  ...   
4702           No  No phone service             DSL                   No  ...   
2361          Yes                No             DSL                   No  ...   

         DeviceProtection          TechSupport          StreamingTV  \
6294  No internet service  No internet service  No internet service   
6201                   No                   No                   No   
4223                   No                   No                   No   
6992                   No                   No                  Yes   
6739                   No                   No                   No   
677                    No                   No                   No   
6444                   No                  Yes                  Yes   
4870                  Yes                   No                   No   
4702                  Yes                   No                  Yes   
2361                   No                   No                   No   

          StreamingMovies        Contract PaperlessBilling  \
6294  No internet service  Month-to-month               No   
6201                   No  Month-to-month               No   
4223                   No  Month-to-month              Yes   
6992                  Yes        One year               No   
6739                  Yes  Month-to-month              Yes   
677                   Yes  Month-to-month              Yes   
6444                  Yes        One year              Yes   
4870                   No        One year              Yes   
4702                  Yes        One year              Yes   
2361                   No  Month-to-month               No   

                  PaymentMethod MonthlyCharges  TotalCharges Churn  
6294               Mailed check          19.70          19.7   Yes  
6201           Electronic check          80.70         374.8    No  
4223               Mailed check          70.30         220.4    No  
6992  Bank transfer (automatic)         100.55       7113.75    No  
6739           Electronic check          55.90         238.5    No  
677            Electronic check          34.70         62.25   Yes  
6444               Mailed check          79.45        2587.7   Yes  
4870  Bank transfer (automatic)          64.40       3035.35    No  
4702           Electronic check          55.45       2966.95    No  
2361               Mailed check          45.95         45.95   Yes  

[10 rows x 21 columns]

In [10]:
#Using a lambda function to replace Fiber Optic and DSL as Yes options
services['InternetService'] = services['InternetService'].apply(lambda a: a.replace('Fiber optic',"Yes,Fiber optic"))\
                                                   .apply(lambda a: a.replace('DSL','Yes,DSL'))

In [11]:
#Melting the dataset to put all the services under one column and their values to show presence or not under another column
services =services.melt(id_vars=['customerID', 'gender','SeniorCitizen','Partner','Dependents','tenure','Contract','PaperlessBilling','PaymentMethod','MonthlyCharges','TotalCharges','Churn'],var_name='Service', value_name='Presence')
#To confirm If the melting was correct
services.sample(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
28348  2656-FMOKZ  Female              1      No         No      15   
19502  8961-QDZZJ  Female              0     Yes        Yes      57   
35914  8338-QIUNR    Male              0     Yes        Yes      72   
22964  8356-WUAOJ  Female              0     Yes         No       1   
57299  9448-REEVD    Male              0     Yes         No       1   
17513  9919-KNPOO  Female              0     Yes         No      21   
39960  6116-RFVHN  Female              0     Yes         No      34   
56929  5192-EBGOV  Female              1      No         No       1   
191    2472-OVKUP    Male              0     Yes         No       6   
34181  6646-VRFOL    Male              1     Yes         No      53   

             Contract PaperlessBilling            PaymentMethod  \
28348  Month-to-month              Yes             Mailed check   
19502        One year              Yes         Electronic check   
35914        Two year              Yes  Credit card (automatic)   
22964  Month-to-month              Yes         Electronic check   
57299  Month-to-month               No  Credit card (automatic)   
17513  Month-to-month              Yes         Electronic check   
39960  Month-to-month              Yes         Electronic check   
56929  Month-to-month              Yes         Electronic check   
191    Month-to-month              Yes         Electronic check   
34181  Month-to-month              Yes         Electronic check   

       MonthlyCharges TotalCharges Churn           Service         Presence  
28348           74.45       1145.7   Yes      OnlineBackup               No  
19502           95.00       5535.8    No   InternetService  Yes,Fiber optic  
35914           66.50       4811.6    No  DeviceProtection              Yes  
22964           45.00           45    No    OnlineSecurity               No  
57299           44.70         44.7   Yes   StreamingMovies               No  
17513           94.30      1948.35    No   InternetService  Yes,Fiber optic  
39960          100.45      3414.65    No  DeviceProtection              Yes  
56929           85.70         85.7   Yes   StreamingMovies               No  
191             91.00          531   Yes      PhoneService              Yes  
34181          103.85       5485.5   Yes      OnlineBackup              Yes

In [12]:
#Checking the values under the column Presence
services['Presence'].value_counts()

No                     24347
Yes                    23685
No internet service     9156
Yes,Fiber optic         3096
Yes,DSL                 2421
No phone service         682
Name: Presence, dtype: int64

In [13]:
#Here, i use loc to get all the rows where the presence is Yes for all the services and the customers churned
services= services.loc[(services['Presence'].isin(['Yes','Yes,Fiber optic','Yes,DSL']))]

In [14]:
services.groupby(['Service','Churn'])['MonthlyCharges'].agg(pd.Series.mode)

Service           Churn
DeviceProtection  No                                             79.2
                  Yes                                   [99.5, 104.4]
InternetService   No                                    [84.8, 89.85]
                  Yes                                            74.4
MultipleLines     No                                               25
                  Yes                                            74.4
OnlineBackup      No                                            89.85
                  Yes                            [94.1, 100.2, 104.4]
OnlineSecurity    No                      [59.75, 79.85, 81.0, 89.85]
                  Yes                                            94.4
PhoneService      No                                            20.05
                  Yes                                            74.4
StreamingMovies   No       [81.0, 99.0, 99.65, 100.3, 100.55, 105.35]
                  Yes                                           93

###### Conclusion
<br>
-> From the above analysis:
<br><br>
*The 3 services contributing to a higher monthly charge and Causing customers to churn are:
<br>
    Device Protection,Online Backup and Streaming TV all having a high monthly charge of 104.4

### b. If the telco was to offer Phone as a standalone service, which type of contract would encourage customer retention?

In [15]:
#For this we check on who have access to PhoneService(Yes), if they churned(No) and what their contract is.
#We then get the most common contract(mode of the column contract)
services.groupby(['Service','Churn'])['Contract'].agg(pd.Series.mode)

Service           Churn
DeviceProtection  No             Two year
                  Yes      Month-to-month
InternetService   No       Month-to-month
                  Yes      Month-to-month
MultipleLines     No             Two year
                  Yes      Month-to-month
OnlineBackup      No             Two year
                  Yes      Month-to-month
OnlineSecurity    No             Two year
                  Yes      Month-to-month
PhoneService      No       Month-to-month
                  Yes      Month-to-month
StreamingMovies   No             Two year
                  Yes      Month-to-month
StreamingTV       No             Two year
                  Yes      Month-to-month
TechSupport       No             Two year
                  Yes      Month-to-month
Name: Contract, dtype: object

### Conclusion
<br>
->From the above analysis:
<br><br>
*A Month-to-month contract would encourage retaintion of customers if the telco was to offer Phone as a standalone service

# Question 3: Payment(Use visualizations in explaining your analysis)

### a. If the company was to streamline all its services into 3 bouquet packages, what monthly prices will be appropriate for the following packages to keep customers from churning:
i. Basic
ii. Plus
iii. Premium

In [16]:
#For this case I would consider the Basic package as the cheapest, Plus package as the middle package and the premium as the most expensive

#I would therefore use the quartiles on the monthly charges column to determine the prices from the customers who were retained(Churn= No)

#First i get all customers who did not churn
services= services.loc[services['Churn']=="No"]

In [17]:
#From the dataset of customers who did not churn we then get the 0.25 quartile of Monthly Charges for Basic for all services, 0.5quartile for Plus for all services and 0.75quartile for Premium for all services
#For each quartile we then get the average 
#The Basic Package
Basic=services.groupby(['Service','Churn'])['MonthlyCharges'].quantile(0.25).mean()
Basic

62.150000000000006

In [18]:
#The Plus Package
Plus=services.groupby(['Service','Churn'])['MonthlyCharges'].quantile(0.5).mean()
Plus

82.87500000000001

In [19]:
#The Premium Package
Premium=services.groupby(['Service','Churn'])['MonthlyCharges'].quantile(0.75).mean()
Premium

98.72083333333333

### Conclusion
<br>
->From the above analysis
<br><br>
*The Basic package would go for 62.15
<br>
*The Plus package would go for 82.88
<br>
*The Premium package would go for 98.72

### b. Should the company strictly go paperless for their monthly billings as a technique of keeping their customers and why?

In [20]:
#In this case we look at the number of customers who were retained(Churn No) and were paying using paperless(Paperless Yes)
services.groupby(['PaperlessBilling'])['Churn'].value_counts()

PaperlessBilling  Churn
No                No        8341
Yes               No       13251
Name: Churn, dtype: int64

### Conclusion

<br>
The company should strictly go paperless as most of their customers who have not churned use PaperlessBilling